Compile on Opensolaris

jhrozek opened the pull request:
This patch adds first part of support for OpenSolaris (tested on Omnios). So far SSL is not supported.

jhrozek committed:
Improve Solaris documentation to include Omnios

jhrozek committed:
Provide compat definition of several OpenLDAP functions

Resolves:
    https://github.com/jhrozek/pam_hbac/issues/79

Some platforms (like Omnios) do not ship with OpenLDAP libraries at all,
but instead only use SunLDAP. And SunLDAP libraries do not provide
ldap_initialize() but only ldap_init(). Also, they don't provide the
modern functions to work with DNs like ldap_str2dn() or ldap_dnfree().
If these are not available, fall back to the deprecated functions like
ldap_explode_dn().

Also provides a unit tests that fakes a deprecated system to run the
existing dnparse-tests just with the compat implementation.

preichl commented on doc/README.Solaris:
```diff
@@ -19,6 +19,15 @@ Building from source
+                developer/build/libtool \
+                developer/pkg-config \
+                developer/build/automake \
+                developer/gcc4.8 \
```
Are you sure? I think it should be developer/gcc48 (without the dot), at least that's how I made it work for me.

preichl commented:
With this patch set I was able to compile, thanks.  I see some warnings:

src/pam_hbac.c: In function ‘pam_hbac_get_items’:
src/pam_hbac.c:120:43: warning: passing argument 3 of ‘pam_get_item’ from incompatible pointer type [-Wincompatible-pointer-types]
     ret = pam_get_item(pamh, PAM_SERVICE, (const void *_) &(pi->pam_service));
                                           ^
In file included from src/pam_hbac_compat.h:30:0,
                 from src/pam_hbac.h:21,
                 from src/pam_hbac.c:30:
/usr/include/security/pam_appl.h:186:1: note: expected ‘void *_’ but argument is of type ‘const void *_’
 pam_get_item(
 ^
src/pam_hbac.c:125:40: warning: passing argument 3 of ‘pam_get_item’ from incompatible pointer type [-Wincompatible-pointer-types]
     ret = pam_get_item(pamh, PAM_USER, (const void *_) &(pi->pam_user));
                                        ^
In file included from src/pam_hbac_compat.h:30:0,
                 from src/pam_hbac.h:21,
                 from src/pam_hbac.c:30:
/usr/include/security/pam_appl.h:186:1: note: expected ‘void *_’ but argument is of type ‘const void *_’
 pam_get_item(
 ^
src/pam_hbac.c:141:39: warning: passing argument 3 of ‘pam_get_item’ from incompatible pointer type [-Wincompatible-pointer-types]
     ret = pam_get_item(pamh, PAM_TTY, (const void *_) &(pi->pam_tty));
                                       ^
In file included from src/pam_hbac_compat.h:30:0,
                 from src/pam_hbac.h:21,
                 from src/pam_hbac.c:30:
/usr/include/security/pam_appl.h:186:1: note: expected ‘void *_’ but argument is of type ‘const void *_’
 pam_get_item(
 ^
src/pam_hbac.c:146:41: warning: passing argument 3 of ‘pam_get_item’ from incompatible pointer type [-Wincompatible-pointer-types]
     ret = pam_get_item(pamh, PAM_RUSER, (const void *_) &(pi->pam_ruser));
                                         ^
In file included from src/pam_hbac_compat.h:30:0,
                 from src/pam_hbac.h:21,
                 from src/pam_hbac.c:30:
/usr/include/security/pam_appl.h:186:1: note: expected ‘void *_’ but argument is of type ‘const void *_’
 pam_get_item(
 ^
src/pam_hbac.c:151:41: warning: passing argument 3 of ‘pam_get_item’ from incompatible pointer type [-Wincompatible-pointer-types]
     ret = pam_get_item(pamh, PAM_RHOST, (const void *_) &(pi->pam_rhost));
                                         ^
In file included from src/pam_hbac_compat.h:30:0,
                 from src/pam_hbac.h:21,
                 from src/pam_hbac.c:30:
/usr/include/security/pam_appl.h:186:1: note: expected ‘void *_’ but argument is of type ‘const void **’
 pam_get_item(
 ^
src/tests/pam_test_client.c:61:5: warning: initialization from incompatible pointer type [-Wincompatible-pointer-types]
     dummy_pam_conv,
     ^
src/tests/pam_test_client.c:61:5: note: (near initialization for ‘conv.conv’)

jhrozek committed:
Silence compilation warnings on Solaris

Resolves:
    https://github.com/jhrozek/pam_hbac/issues/76

On Solaris, pam_get_item returns "char *", but on Linux, the same
interface returns "const char **". Detect this during configure time and
either cast or discard const appropriatelly.

jhrozek commented:
OK, done. There is a WIP patch that also adds a SSL suport. I still need to tweak the manpage though.

jhrozek committed:
config: Rename CA_CERT to SSL_PATH

Since we support NSS databases with certificates in addition to
certificate files, the option must be agnostic as well.

jhrozek committed:
doc: Add README.md

If no README.md was provided in the docs, we would have displayed
README.FreeBSD when a user navigated to the doc/ subdir with a browser
simply because it would be lexicographically first..

jhrozek committed:
LDAP: Add a new option to disable SSL during searches

Add a new undocumented option "SECURE" which defaults to true. If this
option is set to "FALSE", StartTLS or SSL won't be even tried.

jhrozek commented:
@preichl I'm done with this branch, we should merge this patchset and release 1.0. I will run Coverity in the meantime.

jhrozek committed:
Coverity: Make sure we never leak memory in default_hostname

Resolves:
	https://github.com/jhrozek/pam_hbac/issues/51
In particular:
	CID #82323

Coverity was complaining:
 89
   4. noescape: Resource hostname is not freed or pointed-to in gethostname.
 90    ret = gethostname(hostname, HOST_NAME_MAX);
   5. Condition ret == -1, taking true branch
 91    if (ret == -1) {
 92        ret = errno;
   6. Jumping to label done
 93        goto done;
 94    }
 95
 96    /* Make sure that returned string is terminated. */
 97    hostname[HOST_NAME_MAX-1] = '\0';
 98
 99    *_hostname = hostname;
100    ret = 0;
101
102done:
   7. Condition ret != 0, taking false branch
103    if (ret != 0) {
104        free(hostname);
105    }
   CID 82323 (#1 of 1): Resource leak (RESOURCE_LEAK)8. leaked_storage: Variable hostname going out of scope leaks the storage it points to.
106    return ret;
107}

As far as I can see this can happen only if gethostname() fails, but doesn't
set errno. To be on the safe side and have no Coverity warnings, add additional
defensive branch.

preichl commented on src/intgtests/test_pam_hbac.py:
```diff
@@ -333,6 +337,7 @@ def _driver_setup(self):
                                               admin_password,
                                               bind_dn,
                                               bind_pw,
+                                              start_ssl = True,
```
PEP8 complains:
src/intgtests/test_pam_hbac.py:340:56: E251 unexpected spaces around keyword / parameter equals
src/intgtests/test_pam_hbac.py:340:58: E251 unexpected spaces around keyword / parameter equals

I think it would be nice to use same style as for 'insecure',

jhrozek replied:
On Mon, May 23, 2016 at 05:52:40AM -0700, Pavel Reichl wrote:

> > @@ -333,6 +337,7 @@ def _driver_setup(self):
> >                                                admin_password,
> >                                                bind_dn,
> >                                                bind_pw,
> > -                                              start_ssl = True,
> 
> PEP8 complains:
> src/intgtests/test_pam_hbac.py:340:56: E251 unexpected spaces around keyword / parameter equals
> src/intgtests/test_pam_hbac.py:340:58: E251 unexpected spaces around keyword / parameter equals
> 
> I think it would be nice to use same style as for 'insecure',

Yes, done.

preichl commented on src/intgtests/test_pam_hbac.py:
```diff
@@ -375,6 +380,8 @@ def _config_setup(self, host):
         confd['BIND_PW'] = self.driver.bind_pw
         confd['SSL_PATH'] = self.driver.ca_cert
         confd['HOST_NAME'] = host
+        if start_ssl == False:
```
PEP8 complains: src/intgtests/test_pam_hbac.py:383:22: E712 comparison to False should be 'if cond is False:' or 'if not cond:'

But I don't see what's wrong with your version, so I don't really care.

jhrozek replied:
On Mon, May 23, 2016 at 05:55:04AM -0700, Pavel Reichl wrote:

> > @@ -375,6 +380,8 @@ def _config_setup(self, host):
> >          confd['BIND_PW'] = self.driver.bind_pw
> >          confd['SSL_PATH'] = self.driver.ca_cert
> >          confd['HOST_NAME'] = host
> > -        if start_ssl == False:
> 
> PEP8 complains: src/intgtests/test_pam_hbac.py:383:22: E712 comparison to False should be 'if cond is False:' or 'if not cond:'
> 
> But I don't see what's wrong with your version, so I don't really care.

It's just a style issue, Python programmers prefer "is False" over "==
False". I fixed that..

preichl commented on src/pam_hbac_ldap_compat.c:
```diff
@@ -0,0 +1,264 @@
+    }
+
+    /* Skip leading whitespace before the key */
+    rdn_start = exploded_rdn_value;
```
Just a question, should we also skip the space after key and before '=' ?

jhrozek replied:
On Mon, May 23, 2016 at 07:02:50AM -0700, Pavel Reichl wrote:

> > +
> > +#if !defined(HAVE_LDAP_STR2DN) || defined(COMPAT_LDAP_UNIT_TESTS)
> > +static int rdn_keyval(const char *exploded_rdn_value,
> > -                      char *__key, char *__val)
> >   +{
> > -    size_t i;
> > -    char *eq_sign;
> > -    const char *rdn_start;
> > -    const char *val_start;
> >   +
> > -    if (exploded_rdn_value == NULL) {
> > -        return EINVAL;
> > -    }
> >   +
> > -    /\* Skip leading whitespace before the key */
> > -    rdn_start = exploded_rdn_value;
> 
> Just a question, should we also skip the space after key and before '=' ?

Fixed.

preichl commented on src/pam_hbac_ldap_compat.c:
```diff
@@ -0,0 +1,264 @@
+
+    *_key = strndup(rdn_start, eq_sign - rdn_start);
+    *_val = strdup(val_start);
+
```
Should we check on NULL...ENOMEM is not probable, but still...?

jhrozek replied:
On Mon, May 23, 2016 at 07:03:39AM -0700, Pavel Reichl wrote:

> > -        return EINVAL;
> > -    }
> >   +
> > -    val_start = eq_sign+1;
> > -    /\* Skip leading whitespace before the value */
> > -    for (i =0; isspace(val_start[i]); i++);
> >   +
> > -    /\* Empty value? */
> > -    val_start += i;
> > -    if (*val_start == '\0') {
> > -        return EINVAL;
> > -    }
> >   +
> > -    *_key = strndup(rdn_start, eq_sign - rdn_start);
> > -    *_val = strdup(val_start);
> >   +
> 
> Should we check on NULL...ENOMEM is not probable, but still...?

Fixed.

preichl commented on src/pam_hbac_ldap_compat.c:
```diff
@@ -0,0 +1,264 @@
+    ret = rdn_keyval(exploded_rdn[0], &key, &val);
+    ldap_value_free(exploded_rdn);
+    if (ret != 0 || key == NULL || val == NULL) {
+        free(key);
```
Should we free ava here as well?

jhrozek replied:
On Mon, May 23, 2016 at 07:05:10AM -0700, Pavel Reichl wrote:

> > -        return NULL;
> > -    }
> >   +
> > -    ava = malloc(sizeof(LDAPAVA));
> > -    if (ava == NULL) {
> > -        ldap_value_free(exploded_rdn);
> > -        return NULL;
> > -    }
> >   +
> > -    /\* Since the server is IPA, we can ignore anything but the simplest,
> > -     \* single-valued RDNs, we know IPA doesn't use those
> > -     */
> > -    ret = rdn_keyval(exploded_rdn[0], &key, &val);
> > -    ldap_value_free(exploded_rdn);
> > -    if (ret != 0 || key == NULL || val == NULL) {
> > -        free(key);
> 
> Should we free ava here as well?

Fixed.

preichl commented on src/pam_hbac_ldap_compat.c:
```diff
@@ -0,0 +1,264 @@
+    return ava;
+}
+
+static int str2rdn(const char *str_rdn, LDAPRDN *rdn)
```
in rdn_keyval() you used '_' prefix for output param, if you will do any changes in this file you might consider using it here as well...?

jhrozek replied:
On Mon, May 23, 2016 at 07:12:46AM -0700, Pavel Reichl wrote:

> > -    ldap_value_free(exploded_rdn);
> > -    if (ret != 0 || key == NULL || val == NULL) {
> > -        free(key);
> > -        free(val);
> > -        return NULL;
> > -    }
> >   +
> > -    ava->la_attr.bv_val = key;
> > -    ava->la_value.bv_val = val;
> > -    ava->la_attr.bv_len = strlen(key);
> > -    ava->la_value.bv_len = strlen(val);
> >   +
> > -    return ava;
> >   +}
> >   +
> >   +static int str2rdn(const char *str_rdn, LDAPRDN *rdn)
> 
> in rdn_keyval() you used '_' prefix for output param, if you will do any changes in this file you might consider using it here as well...?

Fixed.

preichl commented on src/pam_hbac_ldap_compat.c:
```diff
@@ -0,0 +1,284 @@
+
+    /* Skip whitespace before the '=' sign and after the key */
+    for (rdn_end = eq_sign;
+         rdn_end > rdn_start && isspace(*rdn_end);
```
Is 'rdn_end > rdn_start' necessary? I think that 'rdn_start' must not be space anyway...

Should we also strip trailing white space from value?

jhrozek replied:
On Mon, May 23, 2016 at 07:46:53AM -0700, Pavel Reichl wrote:

> > -    rdn_start = exploded_rdn_value;
> > -    for (i = 0; isspace(rdn_start[i]); i++);
> >   +
> > -    rdn_start += i;
> > -    if (*rdn_start == '\0') {
> > -        return EINVAL;
> > -    }
> >   +
> > -    eq_sign = strchr(rdn_start, '=');
> > -    if (eq_sign == NULL) {
> > -        return EINVAL;
> > -    }
> >   +
> > -    /\* Skip whitespace before the '=' sign and after the key */
> > -    for (rdn_end = eq_sign;
> > -         rdn_end > rdn_start && isspace(*rdn_end);
> 
> Is 'rdn_end > rdn_start' necessary? I think that 'rdn_start' must not be space anyway...

I guess not, fixed.

> Should we also strip trailing white space from value?

Probably yes, fixed.

preichl commented on src/pam_hbac_ldap_compat.c:
```diff
@@ -0,0 +1,284 @@
+        free(new_rdn);
+        return ENOMEM;
+    }
+    new_rdn[1] = NULL;
```
I don't think this line is necessary as calloc() should initialize memory to NULL, but I don't mind to do it explicitly...

jhrozek replied:
On Mon, May 23, 2016 at 07:49:16AM -0700, Pavel Reichl wrote:

> > +
> > +static int str2rdn(const char *str_rdn, LDAPRDN *_rdn)
> > +{
> > -    LDAPAVA **new_rdn;
> >   +
> > -    /\* For single-valued RDNs, this is a noop, for multiple-valued, we just
> > -     \* hope for the best. Because we don't support deny rules, the user
> > -     \* might be not allowed access at worst
> > -     */
> > -    new_rdn = calloc(2, sizeof(LDAPAVA *));
> > -    new_rdn[0] = rdn2ava(str_rdn);
> > -    if (new_rdn[0] == NULL) {
> > -        free(new_rdn);
> > -        return ENOMEM;
> > -    }
> > -    new_rdn[1] = NULL;
> 
> I don't think this line is necessary as calloc() should initialize memory to NULL, but I don't mind to do it explicitly... 

Right, removed.

preichl commented on src/pam_hbac_ldap_compat.c:
```diff
@@ -0,0 +1,284 @@
+#if !defined(HAVE_LDAP_DNFREE) || defined(COMPAT_LDAP_UNIT_TESTS)
+static void free_ava(LDAPAVA *ava)
+{
+    free(ava->la_attr.bv_val);
```
It seems that currently NULL cannot be passed as ava parameter...should we be paranoid?

preichl commented on src/pam_hbac_ldap_compat.c:
```diff
@@ -0,0 +1,284 @@
+
+static void free_rdn(LDAPRDN rdn)
+{
+    free_ava(rdn[0]);
```
Same here?

jhrozek replied:
On Mon, May 23, 2016 at 08:10:37AM -0700, Pavel Reichl wrote:

> > +
> > -    return LDAP_SUCCESS;
> >   +#endif
> >   +}
> >   +
> >   +#if !defined(HAVE_LDAP_DNFREE) || defined(COMPAT_LDAP_UNIT_TESTS)
> >   +static void free_ava(LDAPAVA *ava)
> >   +{
> > -    free(ava->la_attr.bv_val);
> > -    free(ava->la_value.bv_val);
> > -    free(ava);
> >   +}
> >   +
> >   +static void free_rdn(LDAPRDN rdn)
> >   +{
> > -    free_ava(rdn[0]);
> 
> Same here?

Yes, I fixed both. Making free functions a no-op on NULL is just a good
culture.

preichl commented on src/pam_hbac_ldap_compat.c:
```diff
@@ -0,0 +1,297 @@
+         val_end--);
+
+    key = strndup(rdn_start, rdn_end - rdn_start);
+    value = strdup(val_start);
```
Did you forget to change this line - to use val_end and val_start as we do for key?

jhrozek replied:
On Mon, May 23, 2016 at 11:26:31AM -0700, Pavel Reichl wrote:

> > -    /\* Skip leading whitespace before the value */
> > -    for (i =0; isspace(val_start[i]); i++);
> >   +
> > -    /\* Empty value? */
> > -    val_start += i;
> > -    if (*val_start == '\0') {
> > -        return EINVAL;
> > -    }
> >   +
> > -    /\* Strip trailing val whitespace */
> > -    for (val_end = val_start + strlen(val_start);
> > -         isspace(*val_end);
> > -         val_end--);
> >   +
> > -    key = strndup(rdn_start, rdn_end - rdn_start);
> > -    value = strdup(val_start);
> 
> Did you forget to change this line - to use val_end and val_start as we do for key?

Yes, I did. The patches I will push next will also include a test.

preichl commented:
I can see following warning:
src/tests/pam_test_client.c:61:5: warning: initialization from incompatible pointer type [-Wincompatible-pointer-types]
     dummy_pam_conv,
     ^
src/tests/pam_test_client.c:61:5: note: (near initialization for ‘conv.conv’)

preichl commented on doc/README.Solaris:
```diff
@@ -35,6 +44,19 @@ Finally, run the configure script:
             --with-pammoddir=/usr/local/lib/ \
```
should there be  /usr/lib/security as you mentioned a few lines above?

jhrozek replied:
On Mon, May 23, 2016 at 10:14:00PM -0700, Pavel Reichl wrote:

> > @@ -35,6 +44,19 @@ Finally, run the configure script:
> >              --with-pammoddir=/usr/local/lib/ \
> 
> should there be  /usr/lib/security as you mentioned a few lines above?

Thanks, fixed.

jhrozek commented:
On Mon, May 23, 2016 at 11:31:14AM -0700, Pavel Reichl wrote:

> I can see following warning:
> src/tests/pam_test_client.c:61:5: warning: initialization from incompatible pointer type [-Wincompatible-pointer-types]
>      dummy_pam_conv,
>      ^
> src/tests/pam_test_client.c:61:5: note: (near initialization for ‘conv.conv’)

I hope this one would be fixed as well, although I won't have access to
the Omnios environment until the evening.

jhrozek committed:
tests: Fix a compilation warning in pam-test-client.c

jhrozek committed:
man: Fix a typo in the manpage

preichl commented on src/tests/pam_test_client.c:
```diff
@@ -49,10 +49,10 @@ static struct pam_conv conv = {
-                          struct pam_response **response,
-                          void *appdata_ptr)
+int dummy_pam_conv(int num_msg,
+                   const struct pam_message **msgm,
```
removing const fixed the warning for me...if it was just a single \* it would work but there's some magic if it's **...I'm too tired to check now, sorry,,

jhrozek replied:
Yup, seems to work, thanks!

> On 24 May 2016, at 21:28, Pavel Reichl notifications@github.com wrote:
> 
> In src/tests/pam_test_client.c:
> 
> > @@ -49,10 +49,10 @@ static struct pam_conv conv = {
> >      NULL
> >  };
> >  #else
> > -static int dummy_pam_conv(int num_msg,
> > -                          const struct pam_message **msgm,
> > -                          struct pam_response **response,
> > -                          void *appdata_ptr)
> >   +int dummy_pam_conv(int num_msg,
> > -                   const struct pam_message **msgm,
> 
> removing const fixed the warning for me...if it was just a single \* it would work but there's some magic if it's **...I'm too tired to check now, sorry,,
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub

jhrozek committed:
LDAP: Don't specify timeout ourselves

Resolves:
    https://github.com/jhrozek/pam_hbac/issues/88

We tried to be too smart and ended up specifying too huge timeout.
The timeout is supposed to be seconds since the operation, not an
absolute time.

It's really better to not handle the timeout ourselves at all, but rely
on openldap defaults.

jhrozek committed:
doc: Add the ignore_authinfo_unavail to FreeBSD and Solaris docs

Resolves:
    https://github.com/jhrozek/pam_hbac/issues/87

It is useful to include the ignore_authinfo_unavail option in the default
docs, otherwise even local users might be locked out in case the LDAP
server is not reachable.

jhrozek committed:
LDAP: SSL_PATH is a not a required option

Resolves:
    https://github.com/jhrozek/pam_hbac/issues/86

If libldap defaults can be used, just use them and don't force the admin
to specify more parameters than strictly necessary.

jhrozek commented on src/pam_hbac_ldap.c:
```diff
@@ -33,22 +33,17 @@ internal_search(pam_handle_t *pamh,
-    struct timeval tv;
     LDAPMessage *msg;
 
-    ret = gettimeofday(&tv, NULL);
```
This was setting the value too big

jhrozek committed:
doc: Improve RHEL documentation

preichl commented:
I successfully tested this version, thanks for the work and for additional help I required :-).

jhrozek commented:
Thanks for the review!
